src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,406 @@
+        if to_remove == self.current_peer() {
+            let next_index = self.peer_index + 1;
+
+            if next_index == self.peers.len() {
+                GrandpaWarpSync::WaitingForPeers(WaitingForPeers {
+                    peers: self.peers,
+                    genesis_chain_information: self.genesis_chain_information,
+                })
+            } else {
+                GrandpaWarpSync::WarpSyncRequest(Self {
+                    peer_index: next_index,
+                    peers: self.peers,
+                    genesis_chain_information: self.genesis_chain_information,
+                })
+            }
+        } else {
+            let index = self
+                .peers
+                .iter()
+                .position(|peer| peer == &to_remove)
+                .unwrap();
+
+            self.peers.remove(index);
+
+            if index < self.peer_index {
+                self.peer_index -= 1;
+            }
+
+            GrandpaWarpSync::WarpSyncRequest(self)
+        }
+    }
+
+    /// Submit a GrandPa warp sync network response if the request succeeded or
+    /// `None` if it did not.
+    pub async fn handle_response(
+        self,
+        response: Option<Vec<GrandpaWarpSyncResponseFragment>>,
+    ) -> GrandpaWarpSync {
+        let warp_sync_peer = self.current_peer();
+
+        let next_index = self.peer_index + 1;
+
+        match response {
+            Some(response_fragments) => GrandpaWarpSync::Verifier(Verifier {
+                verifier: warp_sync::Verifier::new(
+                    &self.genesis_chain_information,
+                    response_fragments,
+                ),
+                genesis_chain_information: self.genesis_chain_information,
+                warp_sync_peer,
+            }),
+            None if next_index < self.peers.len() => GrandpaWarpSync::WarpSyncRequest(Self {
+                peer_index: next_index,
+                peers: self.peers,
+                genesis_chain_information: self.genesis_chain_information,
+            }),
+            None => GrandpaWarpSync::WaitingForPeers(WaitingForPeers {
+                peers: self.peers,
+                genesis_chain_information: self.genesis_chain_information,
+            }),
+        }
+    }
+}
+
+/// Fetching the parameters for the virtual machine is required to continue.
+pub struct VirtualMachineParamsGet {
+    state: PostVerificationState,
+}
+
+impl VirtualMachineParamsGet {
+    /// Set the code and heappages from storage using the keys `:code` and `:heappages`
+    /// respectively. Also allows setting an execution hint for the virtual machine.
+    pub fn set_virtual_machine_params(
+        self,
+        code: Option<impl AsRef<[u8]>>,
+        heap_pages: Option<impl AsRef<[u8]>>,
+        exec_hint: ExecHint,
+    ) -> GrandpaWarpSync {
+        let (code, heap_pages) = match (code, heap_pages) {
+            (Some(code), Some(heap_pages)) => {
+                use std::convert::TryInto;
```

Comment:
```suggestion
                use core::convert::TryInto as _;
```

Please move at the top of the file as well.